src/Engines/DatabaseEngine.php

```diff
@@ -0,0 +1,342 @@
+            'total' => $this->buildSearchQuery($builder)->toBase()->getCountForPagination(),
+        ];
+    }
+
+    /**
+     * Get the Eloquent models for the given builder.
+     *
+     * @param  \Laravel\Scout\Builder  $builder
+     * @param  int|null  $page
+     * @param  int|null  $perPage
+     * @return \Illuminate\Database\Eloquent\Collection
+     */
+    protected function searchModels(Builder $builder, $page = null, $perPage = null)
+    {
+        $models = $this->buildSearchQuery($builder)->when(
+            ! is_null($page) && ! is_null($perPage),
+            function ($query) use ($page, $perPage) {
+                return $query->forPage($page, $perPage);
+            })->get();
+
+        return count($models) > 0
+                ? $models->filter->shouldBeSearchable()->values()
+                : $models;
+    }
+
+    /**
+     * Initialize / build the search query for the given Scout builder.
+     *
+     * @param  \Laravel\Scout\Builder  $builder
+     * @return \Illuminate\Database\Eloquent\Builder
+     */
+    protected function buildSearchQuery(Builder $builder)
+    {
+        $query = $this->initializeSearchQuery(
+            $builder,
+            $columns = array_keys($builder->model->toSearchableArray()),
+            $this->getPrefixColumns($builder),
+            $this->getFullTextColumns($builder)
+        );
+
+        return $this->constrainForSoftDeletes(
+            $builder, $this->addAdditionalConstraints($builder, $query)
+        );
+    }
+
+    /**
+     * Build the initial text search database query for all relevant columns.
+     *
+     * @param  \Laravel\Scout\Builder  $builder
+     * @param  array  $columns
+     * @param  array  $prefixColumns
+     * @param  array  $fullTextColumns
+     * @return \Illuminate\Database\Eloquent\Builder
+     */
+    protected function initializeSearchQuery(Builder $builder, array $columns, array $prefixColumns = [], array $fullTextColumns = [])
+    {
+        return $builder->model->query()->where(function ($query) use ($builder, $columns, $prefixColumns, $fullTextColumns) {
+            $connectionType = $builder->model->getConnection()->getDriverName();
+
+            $canSearchPrimaryKey = ctype_digit($builder->query) &&
+                                   in_array($builder->model->getKeyType(), ['int', 'integer']) &&
+                                   ($connectionType != 'pgsql' || $builder->query <= PHP_INT_MAX) &&
+                                   in_array($builder->model->getKeyName(), $columns);
+
+            if ($canSearchPrimaryKey) {
+                $query->orWhere($builder->model->getQualifiedKeyName(), $builder->query);
+            }
+
+            $likeOperator = $connectionType == 'pgsql' ? 'ilike' : 'like';
+
+            foreach ($columns as $column) {
+                if (in_array($column, $fullTextColumns)) {
+                    $query->orWhereFullText(
+                        $builder->model->qualifyColumn($column),
```

Comment:
```suggestion
                        $builder->qualifyColumn($column),
```

Available since 5.5 https://github.com/laravel/framework/pull/22577